concert/processes/experiments.py

```diff
@@ -0,0 +1,126 @@
+
+    async def _prepare_run(self, iteration: int):
+        """
+        This function changes whatever should be different between the different experiment
+        executions.
+
+        This could trigger an exchange of the specimen or adjust parameters of the experiment.
+        :param iteration:
+        :type iteration: int
+        """
+        raise NotImplementedError
+
+    async def _get_number_of_iterations(self) -> int:
+        """
+        Should return the total number of iterations.
+        :return: Number of iterations.
+        """
+        raise NotImplementedError
+
+    async def _get_iteration_name(self, iteration: int) -> str:
+        """
+        Function for giving meaningfully names for each experiment execution.
+        Should be overwritten for more complicated naming (e.g. specimen descriptions).
+        :param iteration:
+        :type  iteration: int
+        :return:
+        """
+        return "iteration_%04i" % iteration
+
+    async def _prepare_next_run(self):
+        if await self.get_current_iteration() < await self.get_number_of_iterations() - 1:
+            await self._prepare_run(await self.get_current_iteration() + 1)
+
+    @background
+    @check(source=['standby', 'error'], target="standby")
+    @transition(immediate="running", target="standby")
+    async def run(self):
+        try:
+            await self.prepare()
+            # prepare first iteration
+            self._iteration = 0
+            await self._prepare_run(self._iteration)
+
+            for iteration in range(await self.get_number_of_iterations()):
+                self._iteration = iteration
+                if not self._run_event.is_set():
+                    self._state_value = "paused"
+                await self._run_event.wait()
+                self._state_value = "running"
+
+                self._experiment.walker.descend(await self._get_iteration_name(iteration))
+                exp_run = self._experiment.run()
+                self._experiment.log.info("Sample name: %s" %
+                                          await self._get_iteration_name(iteration))
+
+                # Note: exp_run and self._experiment.ready_to_prepare_next_sample.wait() finish
+                # at the same time, if the user does not implement ready_to_prepare_next_sample.
+                await self._experiment.ready_to_prepare_next_sample.wait()
+                await self._prepare_next_run()
+                await exp_run
+
+        except asyncio.CancelledError:
+            # This is normal, no special state needed -> standby
+            LOG.warning('Experiment director cancelled')
+            self._state_value = 'standby'
+        except Exception as e:
+            # Something bad happened, and we can't know what, so set the state to error
+            LOG.warning(f"Error `{e}' while running experiment director")
+            raise StateError('error', msg=str(e))
+        except KeyboardInterrupt:
```

Comment:
This you don't need to handle, when it's async def it will be caught by session utils.

Reply:
Are you 100% sure? I have the impression, that this is what Daniel saw in his Experiment class.

Reply:
Right, rather keep it in, sorry.